gnovm/pkg/gnolang/type_check.go

```diff
@@ -0,0 +1,882 @@
+				if !ok {
+					return fmt.Errorf(
+						"pointer to non-native type cannot satisfy non-empty native interface; %s doesn't implement %s",
+						pxt.String(),
+						nidt.String())
+				}
+				// if xt has native base, do the naive native.
+				if reflect.PtrTo(nxt.Type).AssignableTo(nidt) {
+					return nil // ok
+				} else {
+					return fmt.Errorf(
+						"cannot use %s as %s",
+						pxt.String(),
+						nidt.String())
+				}
+			} else if xdt, ok := xt.(*DeclaredType); ok {
+				if gno2GoTypeMatches(baseOf(xdt), ndt.Type) {
+					return nil
+				} // not check against native interface
+			} else {
+				return fmt.Errorf(
+					"unexpected type pair: cannot use %s as %s",
+					xt.String(),
+					dt.String())
+			}
+		} else {
+			return errors.New("should not happen")
+		}
+	}
+
+	// case2
+	// Special case if xt or dt is *PointerType to *NativeType,
+	// convert to *NativeType of pointer kind.
+	if pxt, ok := xt.(*PointerType); ok {
+		// *gonative{x} is gonative{*x}
+		//nolint:misspell
+		if enxt, ok := pxt.Elt.(*NativeType); ok {
+			xt = &NativeType{
+				Type: reflect.PtrTo(enxt.Type),
+			}
+		}
+	}
+	if pdt, ok := dt.(*PointerType); ok {
+		// *gonative{x} is gonative{*x}
+		if endt, ok := pdt.Elt.(*NativeType); ok {
+			dt = &NativeType{
+				Type: reflect.PtrTo(endt.Type),
+			}
+		}
+	}
+
+	// Special case of xt or dt is *DeclaredType,
+	// allow implicit conversion unless both are declared.
+	// TODO simplify with .IsNamedType().
+	if dxt, ok := xt.(*DeclaredType); ok {
+		if ddt, ok := dt.(*DeclaredType); ok {
+			// types must match exactly.
+			if !dxt.sealed && !ddt.sealed &&
+				dxt.PkgPath == ddt.PkgPath &&
+				dxt.Name == ddt.Name { // not yet sealed
+				return nil // ok
+			} else if dxt.TypeID() == ddt.TypeID() {
+				return nil // ok
+			} else {
+				return fmt.Errorf(
+					"cannot use %s as %s without explicit conversion",
+					dxt.String(),
+					ddt.String())
+			}
+		} else {
+			// special case if implicitly named primitive type.
+			// TODO simplify with .IsNamedType().
+			if _, ok := dt.(PrimitiveType); ok {
+				return fmt.Errorf(
+					"cannot use %s as %s without explicit conversion",
+					dxt.String(),
+					dt.String())
+			} else {
+				// carry on with baseOf(dxt)
+				xt = dxt.Base // set as base to do the rest check
+			}
+		}
+	} else if ddt, ok := dt.(*DeclaredType); ok {
+		// special case if implicitly named primitive type.
+		// TODO simplify with .IsNamedType().
+		if _, ok := xt.(PrimitiveType); ok { // e.g. 1 == Int(1)
+			if debug {
+				debug.Printf("xt is primitiveType: %v, ddt: %v \n", xt, ddt)
+			}
+			// this is special when dt is the declared type of x
+			if !isUntyped(xt) {
+				return fmt.Errorf(
+					"cannot use %s as %s without explicit conversion",
+					xt.String(),
+					ddt.String())
+			} else { // xt untyped, carry on with check below
+				dt = ddt.Base
+			}
+		} else {
+			dt = ddt.Base
+		}
+	}
+
+	// General cases.
+	switch cdt := dt.(type) {
```

Comment:
What do you think should be the behavior in this block if `xt` is a native type and `dt` is not?

Reply:
it will hit:
https://github.com/gnolang/gno/blob/3ea1b47e2f6ca82f696c7b850dedcc695f30414a/gnovm/pkg/gnolang/preprocess.go#L2710

e.g.
```go
package main

import (
	"fmt"
	"time"
)

func main() {
	var m time.Month
	m = 9

	var m1 int
	m1 = m
	fmt.Println(m1)
}

// Error:
// main/debug/time4_a_native.gno:13: cannot use gonative{time.Month} as int
```